src/levanter/data/audio.py

```diff
@@ -0,0 +1,512 @@
+        )
+
+
+class AudioTextExample(eqx.Module):
+    audio: hax.NamedArray
+    tokens: hax.NamedArray
+    loss_mask: hax.NamedArray
+    attn_mask: AttentionMask | hax.NamedArray = AttentionMask.causal()
+
+    @staticmethod
+    def init(
+        audio: hax.NamedArray,
+        tokens: hax.NamedArray,
+        *,
+        attn_mask: Optional[hax.NamedArray | AttentionMask] = None,
+        loss_mask: Optional[hax.NamedArray] = None,
+        ignore_id: Optional[int] = None,
+    ) -> "AudioTextExample":
+        if tokens.ndim != 1:
+            raise ValueError("tokens must be a 1D array")
+
+        if not jnp.issubdtype(tokens.dtype, jnp.integer):
+            raise ValueError("tokens must be an integer array")
+
+        Pos = tokens.axes[0]
+
+        # don't predict the last token.
+        if loss_mask is None:
+            loss_mask = 1 - hax.nn.one_hot(-1, Pos, dtype=jnp.float32)
+
+        if ignore_id is not None:
+            # we don't compute loss for any tokens matching the ignore index
+            ignore_mask = hax.roll(tokens, -1, Pos) != ignore_id
+            loss_mask = loss_mask * ignore_mask
+
+        return AudioTextExample(audio=audio, tokens=tokens, loss_mask=loss_mask, attn_mask=attn_mask)
+
+
+class AudioTextDataset(ShardableDataset[AudioTextExample]):
+    def __init__(
+        self,
+        dataset: ShardableDataset[AudioTextStorageBatch],
+        TextPos: Axis,
+        AudioPos: hax.AxisSelector,
+        KPos: Axis,
+        fcm_prob: float = 0.0,
+        key: Optional[PRNGKeyArray] = None,
+        ignore_index: Optional[int] = None,
+    ):
+        self.dataset = dataset
+        self.TextPos = TextPos
+        self.AudioPos = AudioPos
+        self.KPos = KPos
+        self.fcm_prob = fcm_prob
+        self.key = key
+        self.ignore_id = ignore_index
+
+        if self.fcm_prob > 0.0 and self.key is None:
+            raise ValueError("must provide key if fcm_prob > 0.0")
+
+    def shard(self, shard_id: int, num_shards: int) -> "AudioTextDataset":
+        return AudioTextDataset(
+            self.dataset.shard(shard_id, num_shards),
+            self.TextPos,
+            self.AudioPos,
+            self.KPos,
+            self.fcm_prob,
+            self.key,
+            self.ignore_id,
+        )
+
+    def __iter__(self) -> Iterator[AudioTextExample]:
+        key = self.key
+        sharding = jax.sharding.SingleDeviceSharding(jax.local_devices(backend="cpu")[0])
+
+        with use_cpu_device():
+
+            @functools.partial(eqx.filter_jit, out_shardings=sharding)
+            def _convert_example(inputs: AudioTextDict, key) -> "AudioTextExample":
+                tokens = hax.named(inputs["input_ids"], self.TextPos)
+                audio_features = hax.named(inputs["input_features"], self.AudioPos)
+                attn_mask = hax.named(inputs["attention_mask"], self.KPos)
+                attn_mask = AttentionMask.explicit(attn_mask)
+
+                if self.fcm_prob > 0:
```

Comment:
probably just rip this out for whisper unless you really want it. would be good to dry this (but probably we want to change where this goes...)